csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
 
 c10::intrusive_ptr<c10d::Work> Broadcast::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
 
   if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, 1);
-    if (params_.dst_bufs.size() == 1) {
-      doLocalCopy(params_.dst_bufs.at(0), params_.src_bufs.at(0));
+    if (params_.is_root_in_mesh) {
+      // Do a local copy and the subsequent broadcast will be in place.
+      doLocalCopy(output_tensor, input_tensor);
     } else {
-      assertBufferCount(params_.dst_bufs, 0);
+      // `output_tensor` isn't allocated for this device.
+      output_tensor = input_tensor;
     }
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
-    assertBufferCount(params_.dst_bufs, 1);
   }
 
   if (params_.team.size() == 1) {
     return nullptr;
   }
 
+  std::vector<at::Tensor> tensors({output_tensor});
   return comm.getBackendForTeam(params_.team, backend)
-      ->broadcast(
-          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
-          {.rootRank = root_relative_index_});
+      ->broadcast(tensors, {.rootRank = root_relative_index_});
 }
 
-Gather::Gather(CommParams params) : Communication(params, "gather") {
-  assertBufferCount(params_.src_bufs, 1);
-}
+Gather::Gather(CommParams params) : Communication(params, "gather") {}
 
 c10::intrusive_ptr<c10d::Work> Gather::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.dst_bufs, params_.team.size());
-    buf_list = {std::move(params_.dst_bufs)};
-  } else {
-    assertBufferCount(params_.dst_bufs, 0);
+
+  if (comm.deviceId() == params_.root && !params_.is_root_in_mesh) {
+    input_tensor = at::empty_like(output_tensor.slice(0, 0, 1));
```

Comment:
Good question! 

1. Allocation is done via PyTorch's cached allocator, so it might not be as expensive as many people thought. At this moment, I've never seen it stick out in real traces. If it indeed sticks out in the future, it will probably be a problem for single-device as well and I'll hope to solve the bigger problem. For example, we could pre-allocate for the max sizes and place some restrictions on dynamic shapes. 
2. With Jie's work on dynamic shapes, shapes won't always be available during the "compile" time. 
3. IIUC, is_root_in_mesh==false in Gather (and several other places in this file) is rare in real workloads. Do you remember why we added them in the first place other than for completeness?

Reply:
> Imagine we execute the same fusion in a loop, then there is no need to reallocate the buffers at each iteration

and the same input shape, yes. I think https://github.com/pytorch/pytorch/issues/124807 (driven by NVIDIA folks) is trying to solve that without requiring too many changes in DL compilers like nvFuser. 

> where do we need to allocate in nvFuser workflow?

There's also allocation of intermediate global buffers: https://github.com/NVIDIA/Fuser/blob/2dbf1d0a36cdd947f2f65094f7713e3fefba0329/csrc/executor.cpp#L1994-L2018

Anyhow, the problem I was trying to say is that nvFuser currently allocates for the exact shape but we will want to reuse the same compiled kernel for different shapes. Because of this gap, if we want the allocation (even though via the caching allocator) to only occur once, we'll have to either reuse the same allocation for different input shapes (e.g. by allocating for the max) or use a different caching granularity for allocation than for kernels. Either way comes with its limitations, but these tradeoffs should definitely be considered.

Reply:
Thanks for the references and explanations!

> and the same input shape, yes. I think https://github.com/pytorch/pytorch/issues/124807 (driven by NVIDIA folks) is trying to solve that without requiring too many changes in DL compilers like nvFuser.

nice

> There's also allocation of intermediate global buffers:

I see, thanks!

> Anyhow, the problem I was trying to say is that nvFuser currently allocates for the exact shape but we will want to reuse the same compiled kernel for different shapes. Because of this gap, if we want the allocation (even though via the caching allocator) to only occur once, we'll have to either reuse the same allocation for different input shapes (e.g. by allocating for the max) or use a different caching granularity for allocation than for kernels. Either way comes with its limitations, but these tradeoffs should definitely be considered.

got it!